### PR TITLE
Add an endpoint for shipping label creation eligibility for Woo mobile

### PR DESCRIPTION
**Pull request by jaclync**
## Description

Endpoint proposal: p91TBi-4pF-p2

### Why

There are quite a few conditions for the mobile client to know whether a store and order are eligible for shipping label creation, it'd be much easier to maintain the logic in the future and avoid double implementation in Android and iOS. In order for mobile to support launching all the features in several milestones (payment method management, customs form, package creation), there are 3 parameters (`can_create_payment_method`, `can_create_customs_form`, `can_create_package`) to support all mobile releases.

- Store country is supported (shared logic with web)
- Store currency is supported (shared logic with web)
- Shipping labels are enabled in account settings (WooCommerce > Settings > Shipping > WooCommerce Shipping) (implemented in web client)
- Order has at least one product that is:
  - Shippable
  - Non-refunded
  - Not already included in an existing non-refunded shipping label
- If the user cannot manage payment methods, there has to be a pre-selected payment method
- If the client doesn't support customs form (`can_create_customs_form` param is set to `false`):
  - Origin and destination address are both in the US
  - Store country is in the US
- If the client doesn't support package creation (`can_create_package` param is set to `false`):
  - At least one package (custom or carrier-provided) has already been created
- If the client doesn't support payment method creation (`can_create_payment_method` param is set to `false`):
  - At least one payment method has been selected

### Changes

My apology for the large amount of diffs 🙇🏻‍♀️ Most of the diffs are in PHP unit tests on the endpoint and shipping label helper function, there are a bunch of test cases especially because of the 3 params to support multiple mobile launches. This is my first time working on PHP and WooCommerce Shipping & Tax, please feel free to suggest **anything**!

- In `WC_Connect_Shipping_Label`:
  - Added a helper function `is_store_eligible_for_shipping_label_creation` that checks whether a store is eligible for shipping label creation: this is shared web in `calculate_should_show_meta_box`
  - Added a helper function `is_order_eligible_for_shipping_label_creation` that checks whether an order is eligible for shipping label creation: the logic differs from the current `calculate_should_show_meta_box` implementation, so I did not share this with web so as not to alter the existing web behavior
- In `WC_REST_Connect_Shipping_Label_Controller`:
  - Added an endpoint `GET /connect/label/{{order_id}}/creation_eligibility` and implemented the logic as in the previous section

### Related issue(s)

https://github.com/woocommerce/woocommerce-ios/issues/2971

### Steps to reproduce & screenshots/GIFs

Please confidence check on the existing behavior in web:
- Make sure the store country is supported (`'US', 'AS', 'PR', 'VI', 'GU', 'MP', 'UM', 'FM', 'MH'`), and the store currency is supported (`USD`)
- Go to an order with at least one physical product --> should see the shipping label block like in the [screenshot](https://user-images.githubusercontent.com/1945542/108670193-6e4b6880-7519-11eb-8179-1d88c00fffe9.png)
- In a separate tab, update the store currency to an unsupported one (e.g. EUR)
- Refresh the order --> should **not** see the shipping label block anymore
- In a separate tab, update the store currency back to a supported one (USD). Then update the store country to an unsupported one (e.g. a country outside of the US)
- Refresh the order --> should **not** see the shipping label block

Other than unit tests and the confidence check above, I'm also manually testing the endpoint with the following scenarios:
- Order with at least one physical product --> eligible
- Order with at least one physical product, and partially refunded --> eligible
- Order with at least one physical product, and fully refunded --> not eligible
- Order with at least one physical product, and all products are already included in pre-existing shipping labels --> not eligible
- Order with no physical products --> not eligible
- Shipping labels are enabled in account settings (WooCommerce > Settings > Shipping > WooCommerce Shipping) --> eligible
- Shipping labels are disabled in account settings (WooCommerce > Settings > Shipping > WooCommerce Shipping) --> not eligible
- When the store country is not in the US but supported and there is no `can_create_customs_form` param --> eligible
- When the user cannot manage payment methods while there is no selected payment method, no `can_create_payment_method` param --> not eligible
- When the user cannot manage payment methods and there is a selected payment method, no `can_create_payment_method` param --> eligible
- When the user can manage payment methods and there is no `can_create_payment_method` param --> eligible
- `can_create_customs_form` param is `false` and all the other requirements are met:
  - When the store country is in the US --> eligible
  - When the store country is not in the US --> not eligible
  - Different countries in origin and destination address --> not eligible
  - Origin address country is in the US but destination address is not --> not eligible
  - Origin and destination address country is both in the US --> eligible
- `can_create_payment_method` param is `false` and all the other requirements are met:
  - The user cannot manage payment methods and there is no selected payment method --> not eligible
  - The user can manage payment methods and there is no selected payment method --> not eligible
  - The user cannot manage payment methods and there is a selected payment method --> eligible
  - The user can manage payment methods and there is a selected payment method --> eligible
- `can_create_package` param is `false` and all the other requirements are met:
  - There are no pre-existing packages (custom or predefined) --> not eligible
  - There is at least one pre-existing package (custom or predefined) --> eligible

### Checklist

- [x] unit tests

- [x] `changelog.txt` entry added

### CODE REVIEW

**Comment by anitaa1990** [comment]
@0nko @hichamboushaba @pmusolino would all 3 of you be able to take a look at this PR? It includes all the eligibility criteria we need in mobile and it would be good to get some eyes from our team before it is approved 😄 Thank you!!

I think we can follow @jaclync's post on `p91TBi-4kU` to setup WCS locally 👍

**Comment by harriswong** [comment]
👀

**Comment by hichamboushaba** [comment]
> @0nko @hichamboushaba @pmusolino would all 3 of you be able to take a look at this PR? It includes all the eligibility criteria we need in mobile and it would be good to get some eyes from our team before it is approved 😄 Thank you!!

I've run some tests locally, and they work as expected, but I had to make a change to properly parse all the `bool` request parameters, for example this line:
```php
$client_can_create_package = isset($request['can_create_package']) ? filter_var($request['can_create_package'], FILTER_VALIDATE_BOOLEAN): true;
```
I didn't write any PHP code for more than 8 years, so maybe I'm just doing something wrong, but my understanding is that the parameters are passed as strings.

I didn't have a chance to test the scenarios involving refunds, I'll try to do it next week.

**Comment by harriswong** [comment]
For the conflict above, `tests/php/test_class-wc-connect-shipping-label.php` is renamed to `tests/php/test-class-wc-connect-shipping-label.php` in `develop` in order to pass phpcs linting. Our phpcs rules requires filename to contain only dashes. 

It maybe a bit confusing to resolve the merge conflict. The easiest way would be to use the `develop`'s one, and copy and paste all your test cases to the bottom.

**Comment by jaclync** [comment]
I've been looking into the unit test failures in `test-class-wc-rest-connect-shipping-label-controller.php` today, and it's from the merge of my other PR https://github.com/Automattic/woocommerce-services/commit/e5bdbffc75106b9b2fa2bb09e7c7722ae74ee212 😅 It's interesting that the tests only fail when running `phpunit` but not `phpunit tests/php/classes/test-class-wc-rest-connect-shipping-label-controller.php`. My guess is from some changes on the settings store, since `enable` and `selected_payment_method_id` account settings aren't getting updated in the failed cases. Will be back for further investigation shortly!

**Comment by jaclync** [comment]
> I've run some tests locally, and they work as expected, but I had to make a change to properly parse all the bool request parameters, for example this line:

@hichamboushaba it's true that passing a boolean parameter value wouldn't work, and I found that passing `0`/`1` works as boolean value in PHP (e.g. `{{site_url}}/wp-json/wc/v1/connect/label/65/creation_eligibility?can_create_customs_form=0` to turn off `can_create_customs_form` flag). @harriswong do you know if this is the recommended way to pass a boolean parameter? Or it's better to pass a string `"true"`/`"false"` value and parse it with `filter_var`?

---

Re merge conflicts and unit test failures: I updated the PR to address these plus addressing phpcs issues and it's ready for review! It was [not calling `super` in test class `setUp`](https://github.com/Automattic/woocommerce-services/pull/2334/commits/3bd712cd413c703de5de753b077c2c5a9dcc008b) that caused the unit test failures 😆

**Comment by harriswong** [comment]
> do you know if this is the recommended way to pass a boolean parameter? Or it's better to pass a string "true"/"false" value and parse it with filter_var?

Good question. `WP_REST_Request`'s `$request` is similar to `$_GET` and `$_POST` that they don't have a type associated with the key-value pair. 

For `GET` request, I'd use `empty()` check. ie. `url?checkBox1=any-value`, `if !empty($request['checkBox1'])` then I would treat it as `true`. If we can't remove the `checkBox1` parameter, then I would have to put a `filter` in to parse the input as suggested above.

For `POST`, I normally pass json data in since `json_decode()` would be able to keep the type.

File: classes/class-wc-connect-shipping-label.php
Comment (harriswong) [inline classes/class-wc-connect-shipping-label.php:353]: An order can have a label generated. If the label is unused, the user can refund it within 30 days. In WCS UI, it is possible that the user print multiple labels on the same order without refunding the previous ones first.

In the code below, as soon as a label is generated (even if it's not used), the `creation_eligibility` end point will return `"is_eligible": false,`. Just want to check with you if this is the expected behaviour.

**Review by harriswong — COMMENTED** [review]
I like the test cases that you added for the end point. The comments explain each test case well and I can understand what each test case was trying to test.

I have some questions. This looks good!

File: tests/php/classes/test-class-wc-rest-connect-shipping-label-controller.php
Comment (jaclync) [reply tests/php/classes/test-class-wc-rest-connect-shipping-label-controller.php:197]: Great catch! These two cases (`test_store_without_existing_packages_is_eligible_for_shipping_label_creation_when_client_can_create_package` and  `test_store_without_existing_packages_is_not_eligible_for_shipping_label_creation_when_client_cannot_create_package`) were missing the setup for empty packages, updated in https://github.com/Automattic/woocommerce-services/pull/2334/commits/0366a822b9dad00bf3c9ef0f62dbfb2c057200da

File: tests/php/classes/test-class-wc-rest-connect-shipping-label-controller.php
Comment (jaclync) [reply tests/php/classes/test-class-wc-rest-connect-shipping-label-controller.php]: Yep, and thanks for the suggestion! mobile client is going to make requests with values like `true`/`false`. In https://github.com/Automattic/woocommerce-services/pull/2334/commits/bc82936913191b72b443394f575dc7b845aeae8e, I updated to pass a string boolean in the test cases (and they failed in the first place) and parse the boolean params with `filter_var( $request[param_name], FILTER_VALIDATE_BOOLEAN )`

File: classes/class-wc-connect-shipping-label.php
Comment (jaclync) [reply classes/class-wc-connect-shipping-label.php:353]: Good question! The current design in mobile is to exclude products that have been in a non-refunded shipping label, but it might be better to match the behavior with the web client. I just asked about this in CGPNUU63E/p1615285371106700-slack - will confirm!

File: tests/php/classes/test-class-wc-rest-connect-shipping-label-controller.php
Comment (harriswong) [reply tests/php/classes/test-class-wc-rest-connect-shipping-label-controller.php]: Nice, Double checked `FILTER_VALIDATE_BOOLEAN` will indeed returns true for "1", "true", "on" and "yes". False otherwise. [ref](https://www.php.net/manual/en/filter.filters.validate.php)

File: classes/class-wc-rest-connect-shipping-label-controller.php
Comment (harriswong) [reply classes/class-wc-rest-connect-shipping-label-controller.php:164]: Thanks for the clarification!

File: classes/class-wc-connect-shipping-label.php
Comment (harriswong) [reply classes/class-wc-connect-shipping-label.php:353]: I dug into the log a bit. I think the button used to be called "Redo shipping" after a label is printed. https://github.com/Automattic/woocommerce-services/commit/81edffe4ee44d0f4fdcb80648692e59d518174ae

The idea of allowing multiple labels printed on the same order might have something to do with redoing fulfillment. I can see this happening when a shipment is missed and they have to reprint a new label.

**Comment by hichamboushaba** [comment]
@anitaa1990 @0nko @pmusolino I pushed a new [commit](16a4c1a) to this branch to update the logic of `can_create_payment_method, the previous logic was checking if we have a `selected_payment_method`, but as we agreed that on M3, the apps will allow selecting different payment methods, just not allowing updating them, we need to check if there are some existing payment methods instead. 

cc @harriswong, and please let me know if my code is not following the conventions, or if there are better ways to do something 🙏 , also I'm holding on fixing the conflict, until we get final decision about allowing re-creating shipping labels.

**Review by 0nko — APPROVED** [review]
LGTM 👍

File: classes/class-wc-rest-connect-shipping-label-controller.php
Comment (harriswong) [inline classes/class-wc-rest-connect-shipping-label-controller.php:267]: Do we need to update "a pre-selected payment method is required." since we allow other payment methods?

**Comment by harriswong** [comment]
@hichamboushaba Thanks! The last 2 commits look good. Regarding the phpcs linting, you can run the following to fix some styles complaint:
```
composer install
./vendor/bin/phpcbf classes/class-wc-rest-connect-shipping-label-controller.php
```

**Comment by hichamboushaba** [comment]
Thanks @harriswong, the linting issues should be fixed now, and I pushed changes to disable the check on non-refunded shipping labels to allow re-creating shipping labels. Ready for another round 🙏

**Comment by harriswong** [comment]
👀

**Review by harriswong — APPROVED** [review]
The last few commits on allowing label purchase on existing labels look good to me. I tested against `http://localhost/wp-json/wc/v1/connect/label/179/creation_eligibility` locally and works as expected. LGTM 👍

**Comment by harriswong** [comment]
Once the pipeline passes, please feel free to use "Squash and merge" to merge.

**Pull request merged**
